### PR TITLE
Increase ruby test coverage

### PR DESCRIPTION
**Pull request by vishrutshah**
This changed adds more tests for Ruby and Azure.Ruby generator.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@vishrutshah__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

**Comment by fearthecowboy** [comment]
@veronicagg can you review this ASAP? (/cc @vishrutshah) I'd like to merge this before the reorg.

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/azure_special_properties_spec.rb [hunk 1]
```diff
@@ -153,7 +153,29 @@
     headers['x-ms-client-request-id'] = @validClientId
     result = @client.xms_client_request_id.get_async(headers).value!
     expect(result.response.status).to eq(200)
-    expect(result.request_id).to eq("123")
+    expect(result.response.headers['x-ms-request-id']).to eq("123")
+  end
+
+  # it 'should not overwrite x-ms-client-request-id' do
```
Comment (veronicagg) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/azure_special_properties_spec.rb]: should this one be removed?

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb [hunk 2]
```diff
@@ -8,221 +8,433 @@
...
+  it 'should succeed for post async no retry' do
+    result = @lros_client.post_async_no_retry_succeeded(@product).value!
+    expect(result.response.status).to eq(200)
+  end
+
+  it 'should fail for post async retry' do
+    expect { @lros_client.post_async_retry_failed(@product).value! }.to raise_error(MsRestAzure::AzureOperationError)
+  end
+
+  it 'should fail for post async no retry canceled' do
```
Comment (veronicagg) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb]: should this one be "should fail for post async retry canceled"?

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb [hunk 3]
```diff
@@ -8,221 +8,433 @@
...
+    expect(result.response.status).to eq(200)
+    expect(result.body.provisioning_state).to eq("Canceled")
+  end
+
+  it 'should succeed for delete 204' do
+    result = @lros_client.delete204succeeded().value!
+    expect(result.response.status).to eq(204)
+  end
+
+  it 'should succeed for delete 200 retry 200' do
```
Comment (veronicagg) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb]: should it be "should succeed for delete 202 retry 200"?

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb [hunk 4]
```diff
@@ -8,221 +8,433 @@
...
+    result = @lros_client.delete202retry200().value!
+    expect(result.response.status).to eq(200)
+  end
+
+  it 'should succeed for delete 200 no retry 204' do
```
Comment (veronicagg) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb]: similar comment as above

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb [hunk 5]
```diff
@@ -8,221 +8,433 @@
...
+
+    test_client = AutoRestLongRunningOperationTestService.new(credentials, base_url)
+    test_client.long_running_operation_retry_timeout = 0
+    @lroretrys_client = test_client.lroretrys
+    @product = Product.new
+    @product.location = "West US"
   end
 
   # Retryable errors
   it 'should retry PUT request on 500 responce' do
```
Comment (veronicagg) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb]: typo in "response"?

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb [hunk 6]
```diff
@@ -8,221 +8,433 @@
...
   it 'should rise error on invalid endpoint received in initial DELETE request' do
-    expect{ @client.lrosads.delete_async_relative_retry_invalid_header().value! }.to raise_exception(MsRestAzure::AzureOperationError)
+    expect{ @lroads_client.delete_async_relative_retry_invalid_header().value! }.to raise_exception(MsRestAzure::AzureOperationError)
   end
 
-  # TODO: Fix flakey test
-  #it 'should rise error on invalid JSON responce in status polling request during DELETE operation' do
-  #  expect{ @client.lrosads.delete_async_relative_retry_invalid_json_polling_async().value! }.to raise_exception(MsRest::DeserializationError)
-  #end
+  it 'should rise error on invalid JSON responce in status polling request during DELETE operation' do
```
Comment (veronicagg) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb]: this test had a comment about being flakey, what changed making it more reliable?

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb [hunk 7]
```diff
@@ -8,221 +8,433 @@
...
+   expect{ @lroads_client.delete_async_relative_retry_invalid_json_polling().value! }.to raise_exception(MsRest::DeserializationError)
+  end
+
+  it 'should serve async PUT operation failed' do
```
Comment (veronicagg) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb]: is this the same as 'should rise error if 400 responce comes in the middle of async PUT operation'?

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/azure_special_properties_spec.rb (hunk 1 above)
Comment (vishrutshah) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/azure_special_properties_spec.rb]: Actually that is a valid test but looks like we have a bung that overrides client-request-id even though we set it to false. I'd prefer to keep it here as when we resolve that we can enable that test as well.  

---
In reply to: [69021567](https://github.com/Azure/autorest/pull/1188#discussion_r69021567) [](ancestors = 69021567)

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb (hunk 2 above)
Comment (vishrutshah) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb]: yup corrected

---
In reply to: [69022094](https://github.com/Azure/autorest/pull/1188#discussion_r69022094) [](ancestors = 69022094)

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb (hunk 3 above)
Comment (vishrutshah) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb]: fixed

---
In reply to: [69023512](https://github.com/Azure/autorest/pull/1188#discussion_r69023512) [](ancestors = 69023512)

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb (hunk 4 above)
Comment (vishrutshah) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb]: corrected

---
In reply to: [69023565](https://github.com/Azure/autorest/pull/1188#discussion_r69023565) [](ancestors = 69023565)

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb (hunk 5 above)
Comment (vishrutshah) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb]: fixed 

---
In reply to: [69024176](https://github.com/Azure/autorest/pull/1188#discussion_r69024176) [](ancestors = 69024176)

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb (hunk 6 above)
Comment (vishrutshah) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb]: I have not changed anything but this did not fail locally on multiple runs. So I assume that other chances that went in regarding client runtime might have fixed it. 

---
In reply to: [69025191](https://github.com/Azure/autorest/pull/1188#discussion_r69025191) [](ancestors = 69025191)

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb [hunk 8]
```diff
@@ -8,221 +8,433 @@
...
+    result = @lroretrys_client.put_async_relative_retry_succeeded(@product).value!
+    expect(result.body.provisioning_state).to eq("Succeeded")
+  end
+
+  it 'should retry on 500 server responce in DELETE request' do
+    result = @lroretrys_client.delete_async_relative_retry_succeeded().value!
+    expect(result.response.status).to eq(200)
+  end
+
+  it 'should serve async POST operation' do
```
Comment (vishrutshah) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb]: yes exactly same. removed it.

---
In reply to: [69024543](https://github.com/Azure/autorest/pull/1188#discussion_r69024543) [](ancestors = 69024543)

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb (hunk 7 above)
Comment (vishrutshah) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb]: yes exactly same. removed it. thanks for catching 

---
In reply to: [69025710](https://github.com/Azure/autorest/pull/1188#discussion_r69025710) [](ancestors = 69025710)

File: AutoRest/Generators/Ruby/Ruby.Tests/RspecTests/validation_spec.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  before(:all) do
+    @base_url = ENV['StubServerURI']
+
+    dummyToken = 'dummy12321343423'
+    @credentials = MsRest::TokenCredentials.new(dummyToken)
+
+    @client = ValidationModule::AutoRestValidationTest.new(@credentials, @base_url)
+  end
+
+  it 'should works with constant in path' do
```
Comment (veronicagg) [inline AutoRest/Generators/Ruby/Ruby.Tests/RspecTests/validation_spec.rb]: typo "work"? same below

Comment (vishrutshah) [inline AutoRest/Generators/Ruby/Ruby.Tests/RspecTests/validation_spec.rb]: corrected

---
In reply to: [69042367](https://github.com/Azure/autorest/pull/1188#discussion_r69042367) [](ancestors = 69042367)

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/azure_special_properties_spec.rb (hunk 1 above)
Comment (veronicagg) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/azure_special_properties_spec.rb]: sounds good!

File: AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb (hunk 6 above)
Comment (veronicagg) [inline AutoRest/Generators/Ruby/Azure.Ruby.Tests/RspecTests/lro_spec.rb]: sounds good!

**Comment by veronicagg** [comment]
@vishrutshah code changes look good to me, build failure needs to be checked. thanks!

**Pull request merged**